### PR TITLE
[WIP] RF: Normalization image/mask preparation

### PR DESCRIPTION
**Pull request by effigies**
Rebased on current master and (I think) using template flow appropriately.

Replaces #236.

### CODE REVIEW

File: niworkflows/interfaces/masks.py [hunk 1]
```diff
@@ -192,6 +194,204 @@ def _post_run_hook(self, runtime):
         return super(ACompCorRPT, self)._post_run_hook(runtime)
 
 
+class PrepareRegistrationImagesInputSpec(BaseInterfaceInputSpec):
+    fixed_file = File(exists=True)
```
Comment (oesteban) [inline niworkflows/interfaces/masks.py]: did you mean
```suggestion
    fixed_file = File(exists=True, mandatory=True)
```
?

File: niworkflows/interfaces/masks.py [hunk 2]
```diff
@@ -192,6 +194,204 @@ def _post_run_hook(self, runtime):
...
+    moving_file = File(exists=True)
```
Comment (oesteban) [inline niworkflows/interfaces/masks.py]:
```suggestion
    moving_file = File(exists=True, mandatory=True)
```

File: niworkflows/interfaces/masks.py [hunk 3]
```diff
@@ -192,6 +194,204 @@ def _post_run_hook(self, runtime):
...
+    fixed_mask = File
```
Comment (oesteban) [inline niworkflows/interfaces/masks.py]: And then these three would be `File(exists=True)`, correct?

File: niworkflows/interfaces/masks.py [hunk 4]
```diff
@@ -192,6 +194,204 @@ def _post_run_hook(self, runtime):
...
+        True, usedefault=True,
+        desc='Set voxels outside the masks to zero thus creating an artificial border that can '
+             'that can drive the registration. Requires reliable and accurate masks. See '
+             'See https://sourceforge.net/p/advants/discussion/840261/thread/27216e69/#c7ba')
+
+
+class PrepareRegistrationImagesOutputSpec(TraitedSpec):
+    fixed_file = File(exists=True)
+    moving_file = File(exists=True)
+    fixed_mask = File
```
Comment (oesteban) [inline niworkflows/interfaces/masks.py]: `exists=True`?

File: niworkflows/interfaces/masks.py [hunk 5]
```diff
@@ -192,6 +194,204 @@ def _post_run_hook(self, runtime):
...
+    moving_mask = File
```
Comment (oesteban) [inline niworkflows/interfaces/masks.py]: `exists=True`?

File: niworkflows/interfaces/masks.py [hunk 6]
```diff
@@ -192,6 +194,204 @@ def _post_run_hook(self, runtime):
...
+        fixed_mask = self.inputs.fixed_mask
+        moving_mask = self.inputs.moving_mask
+        lesion_mask = self.inputs.lesion_mask
+        explicit_masking = self.inputs.explicit_masking
+
+        # Default behavior: pass-through
+        self._results = {
+            'fixed_file': fixed_file,
+            'moving_file': moving_file,
+            }
```
Comment (oesteban) [inline niworkflows/interfaces/masks.py]:
```suggestion
        }
```

File: niworkflows/interfaces/masks.py [hunk 7]
```diff
@@ -192,6 +194,204 @@ def _post_run_hook(self, runtime):
...
+    out_filename = fname_presuffix(out_filename or in_file, suffix='_masked', newpath=cwd)
+
+    img = nb.load(in_file)
+    mask = nb.load(mask_file)
+    try:
+        _check_space(img, mask)
+    except ValueError as err:
+        msg = err.args[0]
+        raise ValueError(msg + 'Files: {}, {}'.format(in_file, mask_file))
+    data = img.get_data() * (mask.get_data() > 0)
```
Comment (oesteban) [inline niworkflows/interfaces/masks.py]: I feel it's safer to keep the original array and just set zeros outside the mask:
```suggestion
data = img.get_data()
data[mask.get_data() <= 0] = 0
```

Comment (effigies) [reply niworkflows/interfaces/masks.py]: I don't understand why, but I can do this, if it matters.

File: niworkflows/interfaces/masks.py [hunk 8]
```diff
@@ -192,6 +194,204 @@ def _post_run_hook(self, runtime):
...
+        lm_img = nb.as_closest_canonical(nb.load(lesion_mask))
+
+        try:
+            _check_space(img, lm_img)
+        except ValueError as err:
+            msg = err.args[0]
+            raise ValueError(msg + 'Files: {}, {}'.format(in_file, lesion_mask))
+
+        if not is_binary(lm_img.get_data()):
+            raise ValueError('`lesion_mask` must be a binary mask')
```
[the remark this answers is not in the document]
Comment (effigies) [reply niworkflows/interfaces/masks.py]: Masks are binary:

https://github.com/bids-standard/bids-specification/pull/109/files#diff-ee997314dc6a77ebab0d882ed7945467R80

**Comment by effigies** [comment]
I ignored the style suggestions; I find the closing brace aligned with content to be more readable, but we can change it if need be. There are a couple other suggestions I'm not convinced on.

But anyway, before this can be merged, it needs to be tested in sMRIPrep. It didn't work last time, so I don't want to supersede @danlurie's work until we're sure this works as well.

**Comment by oesteban** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           

Complexity decreasing per file
==============================
+ niworkflows/interfaces/mni.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096914&bid=16766272)

**Pull request closed without merging**